// authereum/dist/ZkSyncProvider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class ZkSyncProvider {
    constructor(authereum) {
        this.deposit = (params) => __awaiter(this, void 0, void 0, function* () {
            return this.authereum._postMessageToChild({
                method: 'zksyncDeposit',
                params
            });
        });
        this.withdraw = (params) => __awaiter(this, void 0, void 0, function* () {
            return this.authereum._postMessageToChild({
                method: 'zksyncWithdraw',
                params
            });
        });
        this.transfer = (params) => __awaiter(this, void 0, void 0, function* () {
            return this.authereum._postMessageToChild({
                method: 'zksyncTransfer',
                params
            });
        });
        this.authereum = authereum;
    }
}
exports.default = ZkSyncProvider;
